embulk-api/src/test/java/org/embulk/spi/json/TestJsonDouble.java

```diff
@@ -77,4 +90,718 @@ public void testBasicDoubleValue() {
+        assertEquals((short) -1, jsonDouble.shortValue());  // Overflow & fractional
+        assertThrows(ArithmeticException.class, () -> jsonDouble.shortValueExact());
+        assertEquals((int) 2147483647, jsonDouble.intValue());  // Fractional
+        assertThrows(ArithmeticException.class, () -> jsonDouble.intValueExact());
+        assertEquals(31234567890L, jsonDouble.longValue());  // Fractional
+        assertThrows(ArithmeticException.class, () -> jsonDouble.longValueExact());
+        assertEquals(BigInteger.valueOf(31234567890L), jsonDouble.bigIntegerValue());
+        assertThrows(ArithmeticException.class, () -> jsonDouble.bigIntegerValueExact());
+        assertEquals((float) 31234567890.12, jsonDouble.floatValue(), 0.001);
+        assertEquals(31234567890.12, jsonDouble.doubleValue(), 0.001);
+        assertEquals(BigDecimal.valueOf(31234567890.12), jsonDouble.bigDecimalValue());
+        assertEquals("3.123456789012E10", jsonDouble.toJson());
+        assertEquals("3.123456789012E10", jsonDouble.toString());
+        assertEquals(JsonDouble.of(31234567890.12), jsonDouble);
+
+        // JsonDouble#equals must normally reject a fake imitation of JsonDouble.
+        assertFalse(jsonDouble.equals(FakeJsonDouble.of(jsonDouble.doubleValue())));
+    }
+
+    @Test
+    public void testBasicOverLongIntegralValue() {
+        final JsonDouble jsonDouble = JsonDouble.of(9_223_372_036_854_775_807_000.0);
+        assertEquals(JsonValue.EntityType.DOUBLE, jsonDouble.getEntityType());
+        assertFalse(jsonDouble.isJsonNull());
+        assertFalse(jsonDouble.isJsonBoolean());
+        assertFalse(jsonDouble.isJsonLong());
+        assertTrue(jsonDouble.isJsonDouble());
+        assertFalse(jsonDouble.isJsonString());
+        assertFalse(jsonDouble.isJsonArray());
+        assertFalse(jsonDouble.isJsonObject());
+        assertThrows(ClassCastException.class, () -> jsonDouble.asJsonNull());
+        assertThrows(ClassCastException.class, () -> jsonDouble.asJsonBoolean());
+        assertThrows(ClassCastException.class, () -> jsonDouble.asJsonLong());
+        assertEquals(jsonDouble, jsonDouble.asJsonDouble());
+        assertThrows(ClassCastException.class, () -> jsonDouble.asJsonString());
+        assertThrows(ClassCastException.class, () -> jsonDouble.asJsonArray());
+        assertThrows(ClassCastException.class, () -> jsonDouble.asJsonObject());
+        assertEquals(8, jsonDouble.presumeReferenceSizeInBytes());
+        assertTrue(jsonDouble.isIntegral());
+        assertFalse(jsonDouble.isByteValue());
+        assertFalse(jsonDouble.isShortValue());
+        assertFalse(jsonDouble.isIntValue());
+        assertFalse(jsonDouble.isLongValue());
+        assertEquals((byte) -1, jsonDouble.byteValue());  // Overflow
+        assertThrows(ArithmeticException.class, () -> jsonDouble.byteValueExact());
+        assertEquals((short) -1, jsonDouble.shortValue());  // Overflow
```

Comment:
Minor: Do you know any reference why `(double)9_223_372_036_854_775_807_000.0` cast to (short)-1?
It seems that the `(double)9223372036854775807000.0` value encode to `100010001111111010000000000000000000000000000000000000000000000`
I didn't know why it encode to `-1`.

Reply:
Java's primitive conversion has some complex rules. Nice to have a look at: https://docs.oracle.com/javase/specs/jls/se7/html/jls-5.html#jls-5.1.3